drivers/display/display_eadogs164.c

```diff
@@ -0,0 +1,465 @@
+			eadogs0_display_cfg.base_address) != 0) {
+		LOG_DBG("Not able to send command: %d", cmd[1]);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int eadogs164_display_set_brightness(const struct device *dev,
+					const uint8_t brightness)
+{
+	/* No Implementation */
+	return 0;
+}
+
+static int eadogs164_display_set_contrast(const struct device *dev,
+				      const uint8_t contrast)
+{
+	uint8_t cmd[2];
+
+	cmd[0] = CONTROL_BYTE;
+
+	/* Set POWER CONTROL */
+	cmd[1] = EALCD_CMD_POWER_CONTROL;
+	cmd[1] = cmd[1] & (~(0x03));
+	/* clear DB0 and DB1 bits where 0x03 is the mask for C4 and D5 */
+	cmd[1] = cmd[1] | ((contrast & 0x30) >> 4);
+	/* set DB0 and DB1 in destination */
+	if (i2c_write(eadogs0_display_driver.bus, cmd, sizeof(cmd),
+			eadogs0_display_cfg.base_address) != 0) {
+		LOG_DBG("Not able to send command: %d", cmd[1]);
+		return -EINVAL;
+	}	/* now write DB0-DB3 */
+
+	/* Set Contrast */
+	cmd[1] = EALCD_CMD_CONTRAST_SET;
+	cmd[1] = cmd[1] & (~(0x0F));
+	/* clear DB0 and DB3 bits where 0x30 is the mask for DB4 and DB5 */
+	cmd[1] = cmd[1] | (contrast & 0x0F);
+	/* set DB4 and DB5 in destination */
+	if (i2c_write(eadogs0_display_driver.bus, cmd, sizeof(cmd),
+			eadogs0_display_cfg.base_address) != 0) {
+		LOG_DBG("Not able to send command: %d", cmd[1]);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static void eadogs164_display_get_capabilities(const struct device *dev,
+		struct display_capabilities *capabilities)
+{
+	/* No Implementation */
+}
+
+static int eadogs164_display_set_pixel_format(const struct device *dev,
+		const enum display_pixel_format pixel_format)
+{
+	/* No Implementation */
+	return 0;
+}
+
+static int eadogs164_display_set_orientation(const struct device *dev,
+					   const enum display_orientation
+					   orientation)
+{
+	disp_orientation = orientation;
+
+	uint8_t cmd[2];
+
+	cmd[0] = CONTROL_BYTE;
+
+	cmd[1] = EALCD_CMD_FUNCTION_SET;
+	if (i2c_write(eadogs0_display_driver.bus, cmd, sizeof(cmd),
+			eadogs0_display_cfg.base_address) != 0) {
+		LOG_DBG("Not able to send command: %d", cmd[1]);
+		return -EINVAL;
+	}
+	if (disp_orientation == DISPLAY_ORIENTATION_NORMAL) {
+		/* it is top view */
+		cmd[1] = EALCD_CMD_ENTRY_MODE_SET_TOP_VIEW;
+	} else if (disp_orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+		/* it is bottom view */
+		cmd[1] = EALCD_CMD_ENTRY_MODE_SET_BOTTOM_VIEW;
+	} else {
+		LOG_ERR("orientation not supported");
+	}
+	if (i2c_write(eadogs0_display_driver.bus, cmd, sizeof(cmd),
+			eadogs0_display_cfg.base_address) != 0) {
+		LOG_DBG("Not able to send command: %d", cmd[1]);
+		return -EINVAL;
+	}		/* 8 bit data length extension bit, RE=0;IS=1 */
+
+	cmd[1] = EALCD_CMD_FUNCTION_SET_3;
+	if (i2c_write(eadogs0_display_driver.bus, cmd, sizeof(cmd),
+			eadogs0_display_cfg.base_address) != 0) {
+		LOG_DBG("Not able to send command: %d", cmd[1]);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static const struct display_driver_api eadogs164_display_api = {
```

Comment:
I do not think you can use display API with this display controller.

Reply:
@jfischer-phytec-iot, I just checked earlier, I considered grove_lcd api but found display.h more generic as other ssd1306 and ssd16xx are using display.h, besides, the one i am using ssd1803a has more functionalities meeting display.h than other ssd controllers. So the last choice is leaving ssd1803a as standalone and providing public functions for use or defining a new api for such controllers.

Reply:
@vanwinkeljan What do you think about :arrow_up: ? The driver is for SSD1803A, a character (alphanumeric) display controller.

Reply:
@jfischer-phytec-iot In general I have no problem to reuse the display API for character devices as long as they adhere to the API and doe not miss use any API, eg. blanking API should not be misused for blinking (or is this just a typo). Further the device should identify itself as a character device through its capabilities (introduce a dedicated pixel format for character devices?)

The real question is if the users of character based displays can live with the limited API for pixel based displays, only write characters to the display and no cursor control, ...?

Reply:
In my opinion, I think I could rewrite the driver for SSD1803A using display API and address other review remarks and for future work I suggest to extend/generalize display API well adding also other functions of LCD displays (cursor control, scroll etc.). But for now, I can rework this driver considering other review comments and could add public functions for the desired behaviour (if required and not supported by display API). If there are still some concerns I would appreciate to listen.

Reply:
> The real question is if the users of character based displays can live with the limited API for pixel based displays, only write characters to the display and no cursor control, ...?

> future work I suggest to extend/generalize display API well adding also other functions of LCD displays (cursor control, scroll etc.).

I can already imagine that it will cause many false bug reports. I think in this case a custom (possibly derived) API is better.

Reply:
Yes it looks like a custom API would be better in this case, it will make clear to the users that this a completely different kind of display that only supports text. A new API will actually prevent miss usage of the character display device with LVGL or CFB.